core/base/workspace_aliases.hpp

```diff
@@ -0,0 +1,274 @@
+// SPDX-FileCopyrightText: 2011 - 2023 NVIDIA CORPORATION
+// SPDX-FileCopyrightText: 2021 Duane Merrill
+// SPDX-FileCopyrightText: 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause AND Apache-2.0
+
+#ifndef GKO_CORE_BASE_WORKSPACE_ALIASES_HPP_
+#define GKO_CORE_BASE_WORKSPACE_ALIASES_HPP_
+
+
+#include <ginkgo/config.hpp>
+
+
+// Macros for handling different device error return types uniformly
+#if defined(__CUDACC__)
+#define GKO_DEVICE_ERROR_TYPE cudaError_t
+#define GKO_DEVICE_ERROR_INVALID cudaErrorInvalidValue
+#define GKO_DEVICE_NO_ERROR cudaSuccess
+#elif defined(__HIPCC__)
+#define GKO_DEVICE_ERROR_TYPE hipError_t
+#define GKO_DEVICE_ERROR_INVALID hipErrorInvalidValue
+#define GKO_DEVICE_NO_ERROR hipSuccess
+#else
+#define GKO_DEVICE_ERROR_TYPE int
+#define GKO_DEVICE_ERROR_INVALID 1
+#define GKO_DEVICE_NO_ERROR 0
+#endif
+
+
+// This code is a modified version of the code from CCCL
+// (https://github.com/NVIDIA/cccl) (cub/detail/temporary_storage.cuh and
+// cub/temporary_storage.cuh), made available through the Apache-2.0 and BSD-3
+// licenses. See ABOUT-LICENSING.md for more details.
+// The main changes: the namespaces, the object/class
+// names and some simplifications that adhere to Ginkgo's code conventions
+
+
+namespace gko {
+namespace detail {
+
+
+template <int num_allocs, int align_bytes = 8>
+GKO_ATTRIBUTES GKO_INLINE GKO_DEVICE_ERROR_TYPE create_workspace_aliases(
+    void* workspace_ptr, size_t& num_bytes, void* (&allocations)[num_allocs],
+    size_t (&allocation_sizes)[num_allocs])
```

Comment:
```suggestion
    void* workspace_ptr, size_t& num_bytes, void* (&allocations)[num_allocs],
    const size_t (&allocation_sizes)[num_allocs])
```
just ensure it will not be changed accendentally.